### PR TITLE
Support spnego/kerberos authentication for gateway

### PR DESCRIPTION
**Pull request by fifthdegree**
This allows for using kerberos to authenticate to a gateway. It also cleans up support code for authentication (there was some redundancy in nla and gateway)

The benefit is of course better security, but it also adds the potential for using smartcards

Note that for reasons beyond my ability to understand TS Gateway won't accept connections from rpc over http when authenticated using kerberos (schannel is still an option; I think that can be used for smartcard support)

### CODE REVIEW

**Comment by hardening** [comment]
@fifthdegree  you're my hero, I was about to start working on that !

**Comment by hardening** [comment]
@freerdp-bot test please

File: libfreerdp/core/nla.c [hunk 1]
```diff
@@ -136,17 +118,15 @@ struct rdp_nla
...
+
+	rdpAuth* auth;
+	char* pkinitArgs;
+	BYTE certSha1[20];
 };
 
 static BOOL nla_send(rdpNla* nla);
 static int nla_server_recv(rdpNla* nla);
-static void nla_buffer_free(rdpNla* nla);
 static SECURITY_STATUS nla_encrypt_public_key_echo(rdpNla* nla);
```
Comment (akallabeth) [inline libfreerdp/core/nla.c:130]: Does not match implementation

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/core/nla.c (hunk 1 above)
Comment (akallabeth) [reply libfreerdp/core/nla.c:130]: ok, fixed these (and a few other compiler warnings)

**Comment by akallabeth** [comment]
@freerdp-bot test

**Review by akallabeth — APPROVED** [review]

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by hardening** [comment]
@akallabeth sure I'll do a careful review

File: libfreerdp/core/gateway/rts.c [hunk 2]
```diff
@@ -857,9 +857,9 @@ static BOOL rts_read_pdu_bind_nak(wStream* s, rpcconn_bind_nak_hdr_t* ctx)
 	WINPR_ASSERT(s);
 	WINPR_ASSERT(ctx);
 
-	if (!Stream_CheckAndLogRequiredLength(
-	        TAG, s, sizeof(rpcconn_bind_nak_hdr_t) - sizeof(rpcconn_common_hdr_t)))
-		return FALSE;
+	// if (!Stream_CheckAndLogRequiredLength(
```
Comment (hardening) [inline libfreerdp/core/gateway/rts.c:860]: There's something to do here

File: libfreerdp/core/gateway/rts.c [hunk 3]
```diff
@@ -898,9 +898,9 @@ static BOOL rts_read_pdu_fault(wStream* s, rpcconn_fault_hdr_t* ctx)
 	WINPR_ASSERT(s);
 	WINPR_ASSERT(ctx);
 
-	if (!Stream_CheckAndLogRequiredLength(
-	        TAG, s, sizeof(rpcconn_fault_hdr_t) - sizeof(rpcconn_common_hdr_t)))
-		return FALSE;
+	// if (!Stream_CheckAndLogRequiredLength(
```
Comment (hardening) [inline libfreerdp/core/gateway/rts.c:901]: Same here

File: libfreerdp/core/nla.c [hunk 4]
```diff
@@ -97,9 +94,7 @@
  *
  */
 
-#define NLA_PKG_NAME NEGO_SSP_NAME
-
-#define TERMSRV_SPN_PREFIX "TERMSRV/"
+#define NLA_PKG_NAME AUTH_PKG_SPNEGO
```
Comment (hardening) [inline libfreerdp/core/nla.c:97]: is it me or it SPNego the default for NLA ?

**Comment by hardening** [comment]
Ok so first, this is a massive improvement and a very nice mutualization. Still I have general remarks:

* you have changed `nla_decrypt_public_key_echo` and friends to return a `BOOL` instead of a `SECURITY_STATUS`, I find it quite unfortunate because it was giving more informations than just "it worked", "it didn't worked" (especially for OOM cases). I'd be for keeping the original behaviour (except if I missed something important) ;
* I don't like the name `auth` like in `rdpAuth`, `auth.c` or `auth_new` it's way too generic, shall we call that sspiAuth or credSspAuth ? (20% of the time coding, the rest of the time finding names ;) )

File: libfreerdp/core/gateway/rts.c (hunk 2 above)
Comment (fifthdegree) [reply libfreerdp/core/gateway/rts.c]: That wasn't supposed to be part of the commit...

File: libfreerdp/core/nla.c (hunk 4 above)
Comment (fifthdegree) [reply libfreerdp/core/nla.c:97]: Wasn't it always the default?

**Comment by fifthdegree** [comment]
So my reasoning for changing SECURITY_STATUS to BOOL is that `auth` is meant to be an abstraction to authentication. Users of it (like nla) shouldn't really need to be concerned about *why* authentication failed, only that it did. What it does upon failure won't change based on the reason. Of course *we* want to know why authentication failed and to that end `auth` will log the status codes returned.

Regarding the name: honestly, 'auth' seems fine to me, but if you want to change it to something else (whenever you settle on it ;) ) that's fine with me too

**Comment by akallabeth** [comment]
@fifthdegree could you rebase your pr please?
@hardening any more points you wish to be addressed?

**Comment by fifthdegree** [comment]
@akallabeth sure, soon as I get a chance

**Comment by fifthdegree** [comment]
@akallabeth  something got messed up while rebasing I think it has to do with one of your commits wanna take a look?

**Comment by akallabeth** [comment]
@fifthdegree yes, @hardening did some net stuff regarding kerberos and /me did some format strings hardening. looks like the `nla.c` parts need quite the thourough merge as quite a lot changed.

**Comment by akallabeth** [comment]
@fifthdegree clashed with https://github.com/FreeRDP/FreeRDP/pull/8261 looks like you need to manually sort that :/

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by hardening** [comment]
@freerdp-bot can you test the changes please ?

File: libfreerdp/core/gateway/rdg.c [hunk 5]
```diff
@@ -40,6 +41,12 @@
 
 #define TAG FREERDP_TAG("core.gateway.rdg")
 
+#ifdef WITH_SPNEGO
```
Comment (hardening) [inline libfreerdp/core/gateway/rdg.c:44]: Should we refine that `#ifdef` ? Under windows we always have Negotiate, so I feel like `WITH_SPNEGO` is not the right test...

**Review by hardening — APPROVED** [review]
Change the `#ifdef` and we're ready for merge

**Review by hardening — APPROVED** [review]

**Comment by hardening** [comment]
@freerdp-bot can you do the final test please

**Pull request merged**